src/git/git.go

```diff
@@ -53,6 +72,345 @@ func (g *GitVersioning) validate() error {
+	ref, err := g.repo.Head()
+	if err != nil {
+		return fmt.Errorf("error while retrieving the branch pointed to HEAD due to: %s", err.Error())
+	}
+
+	g.branchHead = ref
+
+	return nil
+}
+
+// getCommitHistory aims to retrieve the commit history from a given repository.
+// Returns:
+// 		error: Error whenever something wrong happen.
+func (g *GitVersioning) getCommitHistory() error {
+	defer g.printElapsedTime("GetComitHistory")()
+	g.log.Info("getting commit history")
+	cIter, err := g.repo.Log(&git.LogOptions{From: g.branchHead.Hash(), Order: git.LogOrderCommitterTime})
+	if err != nil {
+		return fmt.Errorf("error while retrieving the commit history  due to: %s", err.Error())
+	}
+
+	var commits []*object.Commit
+	err = cIter.ForEach(func(c *object.Commit) error {
+		commits = append(commits, c)
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error while iterating over the commits due to: %s", err.Error())
+	}
+
+	g.commitHistory = commits
+
+	return nil
+}
+
+// getMostRecentCommit is responsible to get the first commit from a slice of commits sorted in ascending order
+// Returns:
+// 		error: Error whenever something wrong happen.
+func (g *GitVersioning) getMostRecentCommit() error {
+
+	if len(g.commitHistory) == 0 {
+		return fmt.Errorf("no commits found")
+	}
+
+	recentCommit := g.commitHistory[0]
+	for i, commit := range g.commitHistory {
+		if i > 0 {
+			if commit.Author.When.After(g.commitHistory[i-1].Author.When) {
+				recentCommit = commit
+			}
+		}
+	}
+	fmt.Println(recentCommit.Message)
+	g.mostRecentCommit = commitInfo{
+		hash:        recentCommit.Hash.String(),
+		authorName:  recentCommit.Author.Name,
+		authorEmail: recentCommit.Author.Email,
+		message:     recentCommit.Message,
+	}
+	return nil
+}
+
+// getAllTags aims to retrieve all the tag from the repository.
+// Returns:
+// 		err: Error whenever unexpected issues happen.
+func (g *GitVersioning) getAllTags() error {
+	defer g.printElapsedTime("getAllTags")()
+	g.log.Info("getting all tags from repository")
+
+	tagsIter, err := g.repo.Tags()
+	errMessage := "error while retrieving tags from repository due to: %s"
+	if err != nil {
+		return fmt.Errorf(errMessage, err.Error())
+	}
+
+	var tags []object.Tag
+	if err := tagsIter.ForEach(func(ref *plumbing.Reference) error {
+		tags = append(tags, object.Tag{
+			Hash: ref.Hash(),
+			Name: ref.Name().String(),
+		})
+
+		if len(tags) == 0 {
+			return fmt.Errorf(errMessage, "no tags found.")
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf(errMessage, err.Error())
+	}
+
+	g.tagsList = tags
+
+	return nil
+}
+
+// getMostRecentTag aims to get the most recent tag from the repository.
+// Returns:
+// 		err: Error whenever unexpected issues happen.
+func (g *GitVersioning) getMostRecentTag() error {
+	defer g.printElapsedTime("GetMostRecentTag")()
+	g.log.Info("getting most recent tag from repository")
+
+	if len(g.tagsList) == 0 {
+		g.mostRecentTag = "0.0.0"
```

Comment:
do we will generate one release 0.0.0 ? @esequielvirtuoso

Reply:
Nope.
The first one will always be 1.0.0.
This 0.0.0 is necessary to tell other functions that there is no tag yet so they can consider 1.0.0 as the first tag.